src/main/scala/is/hail/io/bgen/BgenRecord.scala

```diff
@@ -31,109 +68,137 @@ class BgenRecordV12(compressed: Boolean, nSamples: Int,
-
-    val nExpectedBytesProbs = (nSamples * (nGenotypes - 1) * nBitsPerProb + 7) / 8
-    assert(reader.length == nExpectedBytesProbs + nSamples + 10, s"Number of uncompressed bytes `${ reader.length }' does not match the expected size `$nExpectedBytesProbs'.")
-
-    rvb.startArray(nSamples) // gs
-    val c0 = Call2.fromUnphasedDiploidGtIndex(0)
-    val c1 = Call2.fromUnphasedDiploidGtIndex(1)
-    val c2 = Call2.fromUnphasedDiploidGtIndex(2)
-
-    i = 0
-    while (i < nSamples) {
-      val sampleMissing = (a(8 + i) & 0x80) != 0
-      if (sampleMissing)
-        rvb.setMissing()
-      else {
-        rvb.startStruct() // g
-
-        val off = nSamples + 10 + 2 * i
-        val d0 = a(off) & 0xff
-        val d1 = a(off + 1) & 0xff
-        val d2 = 255 - d0 - d1
-
-        if (includeGT) {
-          if (d0 > d1) {
-            if (d0 > d2)
-              rvb.addInt(c0)
-            else if (d2 > d0)
-              rvb.addInt(c2)
-            else {
-              // d0 == d2
-              rvb.setMissing()
+      val reader = new ByteArrayReader(bytes)
+
+      val nRow = reader.readInt()
+      if (nRow != nSamples)
+        fatal(s"$nRow $nSamples")
+
+      val nAlleles = reader.readShort()
+      if (nAlleles != expectedNumAlleles)
+        fatal(s"Value for `nAlleles' in genotype probability data storage is not equal to value in variant identifying data. Expected $expectedNumAlleles but found $nAlleles.")
+      if (nAlleles != 2)
+        fatal(s"Only biallelic variants supported, found variant with $nAlleles")
+
+      val minPloidy = reader.read()
+      val maxPloidy = reader.read()
+
+      if (minPloidy != 2 || maxPloidy != 2)
+        fatal(s"Hail only supports diploid genotypes. Found min ploidy equals `$minPloidy' and max ploidy equals `$maxPloidy'.")
+
+      reader.skipBytes(nSamples)
+
+      val phase = reader.read()
+      if (!(phase == 0 || phase == 1))
+        fatal(s"Value for phase must be 0 or 1. Found $phase.")
+      val isPhased = phase == 1
+
+      if (isPhased)
+        fatal("Hail does not support phased genotypes.")
+
+      val nBitsPerProb = reader.read()
+      if (!(nBitsPerProb >= 1 && nBitsPerProb <= 32))
+        fatal(s"Value for nBits must be between 1 and 32 inclusive. Found $nBitsPerProb.")
+      if (nBitsPerProb != 8)
+        fatal(s"Only 8-bit probabilities supported, found $nBitsPerProb")
+
+      rvb.startArray(nSamples) // gs
+      if (nBitsPerProb == 8 && nAlleles == 2) {
+        val c0 = Call2.fromUnphasedDiploidGtIndex(0)
+        val c1 = Call2.fromUnphasedDiploidGtIndex(1)
+        val c2 = Call2.fromUnphasedDiploidGtIndex(2)
+
+        var i = 0
+        while (i < nSamples) {
+          val off = nSamples + 10 + 2 * i
+          val d0 = bytes(off) & 0xff
+          val d1 = bytes(off + 1) & 0xff
+
+          if ((bytes(8 + i) & 0x3f) != 2)
+            fatal(s"Ploidy value must equal to 2. Found ${(bytes(8 + i) & 0x3f)}")
+
+          if ((bytes(8 + i) & 0x80) != 0)
```

Comment:
Why are you reading the bytes above if the sample is missing? I think it was better in the original order.